### ISSUE TITLE
keystore password is required when adding compute into keylime_verifier with keylime_tenant

### ISSUE THREAD

**Issue by zkolpet** [keylime/keylime#1135]
Is your issue a feature request? If so, please raise it as an [enhancement](https://github.com/keylime/enhancements)

# Environment
 * OS / version: Ub22.04
 * Processor architecture: N/A
 * TPM Manufacturer: N/A
 * Keylime version: v6.5.1

# Description 
Keylime CA component requires a password set in its config (ca.conf)

# The keystore password
# This password is used to protect the generated CA private key.
password = some_password

if I don't set any password here, the keylime_verifier service will not start...and then if I set some password in this config, every time I am adding some compute/agent into verifier (via keylime_tenant) to be attestated, I am asked for this password from ca.conf:

This prompt shows up:
Please enter the password to decrypt your keystore:

would be great to have a way around this, so I can add compute/agent into verifier without being prompted for this password?

## Expected behavior vs. actual behavior 
No keystore password is needed when adding compute/agent into keylime_verifier via keylime_tenant tool.

## Steps to reproduce problem
1. Start keylime server
2. Start keylime agent
3. Add compute/agent into verifier using keylime_tenant tool

# Relevant logs

Attach any relevant log files that can help to debug your issue.

**Comment by maugustosilva** [comment]
@THS-on @ansasaki @mpeters I would like to be able to have the password read from a file (e.g., `password = /path/to/file`) or from the an environment variable (e.g., `password = <name of an environment variable>`). 

I already have code to do exactly this `cli/options.py` (`extract_password`). Does anyone see this approach as unacceptable? Please note final goal is to have **no** passwords on keylime configuration files.

**Comment by THS-on** [comment]
I think that the detection if there is an environment variable or a path has its issues. For example if an attacker can delete the file then they now can force the password to be the path and the precedence is not obvious.

A cleaner solution in my opinion is to allow any option to be overwritten by an environment variable with a fixed naming scheme e.g. `KEYLIME_COMPONENT_OPTION`. This has the advantage that it is generic, can be easily documented and we can define that environment options take always a precedence over the config file.

**Comment by mpeters** [comment]
While I do like @THS-on 's suggestion to use a generic env-overrides-conf-file approach, maybe it's still possible to do this separately at first if that is too big of a lift. But I also agree that having too many options is probably not the best and passwords stored in plain text files gives me the willies. So, maybe we allow a value of `environment` as the password to pull the value from the `KEYLIME_CA_PASSWORD` environment variable? And then when we can override all keylime conf options with env vars this will just continue to work anyway. Thoughts?

**Comment by maugustosilva** [comment]
> I think that the detection if there is an environment variable or a path has its issues. For example if an attacker can delete the file then they now can force the password to be the path and the precedence is not obvious.
> 
> A cleaner solution in my opinion is to allow any option to be overwritten by an environment variable with a fixed naming scheme e.g. `KEYLIME_COMPONENT_OPTION`. This has the advantage that it is generic, can be easily documented and we can define that environment options take always a precedence over the config file.

Our first priority should be removing the need for plaintext passwords from the configuration file, but you indeed have a valid point about my (admittedly crude) detection mechanism for "implict" source for the password.

In order to address it, I propose the that we prefix the contents of the password with something like `file:/path/to/file` or `env:<name of the env var>`, `plaintext:<plaintext password>` or simply `<plaintext password>` (implicitly assuming it is a plaintext). Would that be acceptable for me to move forward with the implementation?

**Comment by THS-on** [comment]
> In order to address it, I propose the that we prefix the contents of the password with something like file:/path/to/file or env:<name of the env var>, plaintext:<plaintext password> or simply <plaintext password> (implicitly assuming it is a plaintext). Would that be acceptable for me to move forward with the implementation?

The issue of adding something like this now is that we need to maintain that and I would like to keep our configuration as simple as possible.

> I believe that having the ability to override anything on the configuration files with values from environment variables with a fixed naming scheme is a welcome addition, and can immediately envision several scenarios where the lives of the devops teams will be made easier by it. That being said, it is a much larger undertake, one that I am not willing to go through at moment, while being cognizant of the need for removing plaintext passwords from config files. Hence, the compromise solution I propose here.

I actually think it is pretty straight forward to implement the environment variable way, because it is basically just adding a check to the `get*` fucntions in `config.py` if there exists the environment variable and then outputting the data.

**Comment by maugustosilva** [comment]
@THS-on we still need to address a case were the password is stored on a file (which is a very common case for our production environments). It looks to me that passwords will continue to be a special case, and thus I would argue that paying the price for maintaining this extra bit of (optional) configuration is worth the effort.

**Comment by THS-on** [comment]
@maugustosilva is there a reason that you cannot use something like EnvironmentFile from systemd (https://www.freedesktop.org/software/systemd/man/systemd.exec.html#EnvironmentFile=) or something similar?

**Comment by aplanas** [comment]
I personally would not use clear text files. I know that eventually I will make a mistake and make the content available.

For the same reason I would not make use of environment variables. I know that I will put it in a shell script, or in a docker file without the correct permissions. Also I do not have a full understanding of how to protect (if required) the `/proc/<PID>/environ` file associated with the service.

I vote to use systemd credentials: https://systemd.io/CREDENTIALS/ and systemd-creds to manage them. In this model we can put the password encrypted, and make is accessible to the service by a file that it is only accessible by this service.

Keylime can read this file with the secret password in it (in clean form) from a location that will be communicated via an environment variable, using `LoadCredencialEncrypted=` or `SetCredentialEncrypted=` systemd unit parameter.

**Comment by THS-on** [comment]
> For the same reason I would not make use of environment variables. I know that I will put it in a shell script, or in a docker file without the correct permissions. Also I do not have a full understanding of how to protect (if required) the /proc/<PID>/environ file associated with the service.

Good point, but in general having a way of overriding options via environment variable is a nice feature and we would get that option for free if we implement it. 

> I vote to use systemd credentials: https://systemd.io/CREDENTIALS/ and systemd-creds to manage them. In this model we can put the password encrypted, and make is accessible to the service by a file that it is only accessible by this service.

That looks like the way to go, for loading passwords from other locations. Thanks for suggesting it, because I haven't heard of that feature before.

**Comment by aplanas** [comment]
> in general having a way of overriding options via environment variable is a nice feature and we would get that option for free if we implement it.

Yes.

> That looks like the way to go for loading passwords from other locations.

Is a general way of providing secrets to services. A nice feature is that it can use the local TPM2 for sealing / unsealing, closing a nice circle here : )

**Comment by maugustosilva** [comment]
@aplanas thanks for suggesting this, I will take a look. One of the pain points is the need type a password whenever the CLI `keylime_tenant` is invoked with certain options (e.g., `--include` and `--cert`) are used. I will read more, but my first reaction is that I would have to wrap call to the CLI in a systemd service in order to be able to make use of what you suggest, right?

@THS-on just to summarize: we are on "violent agreement" on the need for the ability of overriding options on the configuration files via environment variables.

The only point left to discuss is how to get rid of plaintext passwords on the configuration files. 

There are several other tools, and several other teams which rely on passwords on files on the production  environment (think K8s Secrets with and without Hashicorp vault) and I believe that giving them a way to seamlessly (without having to resort to EnvironmentFile from systemd) use it with keylime is valuable. However, I am open to arguments to the contrary.

**Comment by galmasi** [comment]
A note of caution about assuming systemd to be available on the system. I just checked OpenBMC and to my amazement it does seem to support systemd, so I don't really have a "killer argument against", but I feel like we may be creating a system/packaging dependency we should be careful to at the very least keep *optional*.

**Comment by THS-on** [comment]
> The only point left to discuss is how to get rid of plaintext passwords on the configuration files.

I don't think we can get fully rid of them, but we can remove them from the default config. I suggest the following steps:
* Add support for overriding options using environment variables. (Mostly done already in https://github.com/keylime/keylime/pull/1274)
* Implement a systemd CREDENTIALS integration with a special keyword. I don't think we need to introduce the flexibility with something like "credentials:CREDENTIAL_NAME", we can just always use the options key.
* Switch the default configuration to contain that keyword by default and fail otherwise.

**Comment by aplanas** [comment]
I make some improvements in `systemd-creds`, so is more safe to use it without root:

https://github.com/systemd/systemd/pull/26054

The idea, once #1274 from @THS-on lands, is to use something like:

```bash
# Setup for a user a Keylime credential (only needed one time)
export SYSTEMD_CREDENTIAL_SECRET=$HOME/.config/systemd/credential.secret
systemd-creds setup
systemd-ask-password -n | systemd-creds encrypt - keylime.cred

# I can wrap this in a shell script, or create an alias
KEYLIME_TENANT_PASSWORD=$(systemd-creds decrypt keylime.cred) keylime_tenant ...
```

**Comment by maugustosilva** [comment]
Currently, this environment variable (should be `KEYLIME_CA_PASSWORD` , no?) is **not** used neither by `keylime_ca` nor by `keylime_tenant`. 
I assume we are in agreement that a PR that effectively makes use of it (function `read_private` on `ca_util.py`) will also be needed. 

To be clear: this will be used whenever one wants to ship files to a `agent`, which requires a combinations of options `--include` and `--cert` on `keylime_tenant`

**Comment by THS-on** [comment]
Yes, if I see that correctly we still need to add that.
